File: sync/sync_project.py
#!/usr/bin/env python3

import yaml
import os
import sys
import argparse

def create_files_from_yaml(data, base_path=None):
    summaries = []

    for file in data['files']:
        path = file['path']
        if base_path and not path.startswith(base_path):
            continue
        content = file['content']
        os.makedirs(os.path.dirname(path), exist_ok=True)
        with open(path, 'w') as f:
            f.write(content)
        summaries.append(f"{path}: {content.splitlines()[0]}")

    return summaries

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Sync project files from YAML input.")
    parser.add_argument('--path', type=str, help='Base path to narrow the scope of files processed.')
    args = parser.parse_args()

    # Read YAML content from standard input
    yaml_content = sys.stdin.read()
    data = yaml.safe_load(yaml_content)

    summaries = create_files_from_yaml(data, args.path)

    for summary in summaries:
        print(summary)

File: sync/sync/sync_project.rb
#!/usr/bin/env ruby

require 'yaml'
require 'fileutils'
require 'optparse'

def create_files_from_yaml(data, base_path = nil)
  summaries = []

  data['files'].each do |file|
    path = file['path']
    next if base_path && !path.start_with?(base_path)

    content = file['content']
    dirname = File.dirname(path)
    FileUtils.mkdir_p(dirname) unless File.directory?(dirname)

    File.open(path, 'w') { |f| f.write(content) }
    summaries << "#{path}: #{content.lines.first.strip}"
  end

  summaries
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: sync_project.rb [options]"
  opts.on("--path PATH", "Base path to narrow the scope of files processed") do |v|
    options[:path] = v
  end
end.parse!

yaml_content = STDIN.read
data = YAML.safe_load(yaml_content)

summaries = create_files_from_yaml(data, options[:path])
summaries.each { |summary| puts summary }
